File: .claude/worktrees/distracted-torvalds-4714d9/services/enrichment/workers/tidal.js
'use strict';
// ── Tidal worker ──────────────────────────────────────────────────────────────
// Hergebruikt de bestaande Tidarr-connectie om Tidal-metadata op te halen.
// Geen directe Tidal API-calls — gaat via de lokale Tidarr instantie.

const RATE_INTERVAL = 3000; // 20/min max

class TidalWorker {
  constructor(db, log, { tidarrUrl, tidarrApiKey } = {}) {
    this.db         = db;
    this.log        = log.child ? log.child({ worker: 'tidal' }) : log;
    this.tidarrUrl  = tidarrUrl  || process.env.TIDARR_URL || 'http://localhost:8484';
    this.apiKey     = tidarrApiKey || process.env.TIDARR_API_KEY || null;
    this._lastCall  = 0;
  }

  get isConfigured() {
    return !!this.tidarrUrl;
  }

  async _rateLimit() {
    const wait = Math.max(0, RATE_INTERVAL - (Date.now() - this._lastCall));
    if (wait > 0) await new Promise(r => setTimeout(r, wait));
    this._lastCall = Date.now();
  }

  _headers() {
    const h = { 'Accept': 'application/json', 'Content-Type': 'application/json' };
    if (this.apiKey) h['X-Api-Key'] = this.apiKey;
    return h;
  }

  async _get(path) {
    await this._rateLimit();
    const res = await fetch(`${this.tidarrUrl}${path}`, {
      headers: this._headers(),
      signal:  AbortSignal.timeout(10_000),
    });
    if (res.status === 404) return null;
    if (!res.ok) throw new Error(`Tidarr HTTP ${res.status}`);
    return res.json();
  }

  async process(entity) {
    if (!this.isConfigured) return { ok: false, error: 'Tidarr URL not configured' };
    try {
      const data = await this.search(entity.entity_name, entity.entity_type);
      if (!data) return { ok: false, error: 'No results found' };
      return { ok: true, data };
    } catch (err) {
      this.log.warn({ err: err.message, entity: entity.entity_name }, 'Tidal/Tidarr fetch failed');
      return { ok: false, error: err.message };
    }
  }

  async search(name, entityType = 'artist') {
    try {
      // Tidarr search endpoint
      const encoded = encodeURIComponent(name);
      const type    = entityType === 'album' ? 'albums' : entityType === 'track' ? 'tracks' : 'artists';
      const data    = await this._get(`/api/search?query=${encoded}&type=${type}`);
      if (!data || !data.items?.length) return null;

      const norm  = s => (s || '').toLowerCase().trim();
      const exact = data.items.find(i =>
        norm(i.name || i.title || i.artistName) === norm(name)
      );
      const best = exact || data.items[0];

      return this._mapItem(best, entityType);
    } catch (err) {
      // Tidarr is optioneel — geef null terug bij connectieproblemen
      if (err.message.includes('fetch') || err.message.includes('ECONNREFUSED')) {
        this.log.debug({ name }, 'Tidarr not reachable, skipping Tidal enrichment');
        return null;
      }
      throw err;
    }
  }

  _mapItem(item, entityType) {
    if (entityType === 'artist') {
      return {
        tidalId:      item.id || null,
        name:         item.name || null,
        popularity:   item.popularity || null,
        picture:      item.picture || null,
        url:          item.url || null,
        source:       'tidal',
        fetchedAt:    Date.now(),
      };
    } else if (entityType === 'album') {
      return {
        tidalId:      item.id || null,
        title:        item.title || null,
        artistName:   item.artist?.name || null,
        releaseDate:  item.releaseDate || null,
        duration:     item.duration || null,
        numberOfTracks: item.numberOfTracks || null,
        audioQuality: item.audioQuality || null,
        cover:        item.cover || null,
        explicitLyrics: item.explicitLyrics || false,
        url:          item.url || null,
        source:       'tidal',
        fetchedAt:    Date.now(),
      };
    }
    return {
      tidalId:      item.id || null,
      title:        item.title || null,
      artistName:   item.artist?.name || null,
      albumTitle:   item.album?.title || null,
      duration:     item.duration || null,
      trackNumber:  item.trackNumber || null,
      audioQuality: item.audioQuality || null,
      isrc:         item.isrc || null,
      url:          item.url || null,
      source:       'tidal',
      fetchedAt:    Date.now(),
    };
  }
}

module.exports = { TidalWorker };

File: core/workers/tidal.py
"""
tidal.py — Tidal enrichment worker via Tidarr API proxy.

Geen directe Tidal API-calls — gaat via de lokale Tidarr instantie.
Rate limit: conservatief 1 call per 3 sec (20/min max).
"""

import logging
import os
import time
from typing import Any, Optional

import requests

from core.workers.base import BaseWorker

RATE_INTERVAL   = 3.0
REQUEST_TIMEOUT = 10


class TidalWorker(BaseWorker):
    source = "tidal"

    def __init__(self, database, logger: logging.Logger,
                 tidarr_url: Optional[str] = None,
                 tidarr_api_key: Optional[str] = None):
        super().__init__(database, logger)
        self._url       = tidarr_url or os.environ.get("TIDARR_URL", "http://localhost:8484")
        self._api_key   = tidarr_api_key or os.environ.get("TIDARR_API_KEY")
        self._last_call = 0.0
        self._session   = requests.Session()
        self._session.headers.update({"Accept": "application/json", "Content-Type": "application/json"})
        if self._api_key:
            self._session.headers["X-Api-Key"] = self._api_key

    @property
    def is_configured(self) -> bool:
        return bool(self._url)

    def _rate_limit(self) -> None:
        wait = max(0.0, RATE_INTERVAL - (time.time() - self._last_call))
        if wait > 0:
            time.sleep(wait)
        self._last_call = time.time()

    def _get(self, path: str) -> Optional[dict]:
        self._rate_limit()
        resp = self._session.get(f"{self._url}{path}", timeout=REQUEST_TIMEOUT)
        if resp.status_code == 404:
            return None
        resp.raise_for_status()
        return resp.json()

    def process(self, item: dict) -> dict[str, Any]:
        if not self.is_configured:
            return {"ok": False, "error": "Tidarr URL niet geconfigureerd"}
        try:
            data = self._search(item["entity_name"], item["entity_type"])
            if not data:
                return {"ok": False, "error": "No results found"}
            return {"ok": True, "data": data}
        except Exception as exc:
            self.log.warning(f"Tidal/Tidarr worker mislukt voor '{item.get('entity_name')}': {exc}")
            return {"ok": False, "error": str(exc)}

    def _search(self, name: str, entity_type: str = "artist") -> Optional[dict]:
        from urllib.parse import quote
        try:
            search_type = (
                "albums"  if entity_type == "album" else
                "tracks"  if entity_type == "track" else
                "artists"
            )
            data = self._get(f"/api/search?query={quote(name)}&type={search_type}")
            if not data or not data.get("items"):
                return None

            norm  = lambda s: (s or "").lower().strip()
            items = data["items"]
            exact = next(
                (i for i in items if norm(i.get("name") or i.get("title") or i.get("artistName", "")) == norm(name)),
                None,
            )
            best = exact or items[0]
            return self._map_item(best, entity_type)

        except requests.exceptions.ConnectionError:
            self.log.debug(f"Tidarr niet bereikbaar, Tidal enrichment overgeslagen voor '{name}'")
            return None

    def _map_item(self, item: dict, entity_type: str) -> dict:
        if entity_type == "artist":
            return {
                "tidalId":    item.get("id"),
                "name":       item.get("name"),
                "popularity": item.get("popularity"),
                "picture":    item.get("picture"),
                "url":        item.get("url"),
                "source":     "tidal",
                "fetchedAt":  int(time.time() * 1000),
            }
        if entity_type == "album":
            return {
                "tidalId":          item.get("id"),
                "title":            item.get("title"),
                "artistName":       (item.get("artist") or {}).get("name"),
                "releaseDate":      item.get("releaseDate"),
                "duration":         item.get("duration"),
                "numberOfTracks":   item.get("numberOfTracks"),
                "audioQuality":     item.get("audioQuality"),
                "cover":            item.get("cover"),
                "explicitLyrics":   item.get("explicitLyrics", False),
                "url":              item.get("url"),
                "source":           "tidal",
                "fetchedAt":        int(time.time() * 1000),
            }
        # track
        return {
            "tidalId":     item.get("id"),
            "title":       item.get("title"),
            "artistName":  (item.get("artist") or {}).get("name"),
            "albumTitle":  (item.get("album") or {}).get("title"),
            "duration":    item.get("duration"),
            "trackNumber": item.get("trackNumber"),
            "audioQuality": item.get("audioQuality"),
            "isrc":        item.get("isrc"),
            "url":         item.get("url"),
            "source":      "tidal",
            "fetchedAt":   int(time.time() * 1000),
        }
